docs/syntax-key-destructuring.md

```diff
@@ -0,0 +1,182 @@
+
+```lua
+local position = props.position
+local style = props.style
+-- etc...
+```
+
+...whereas in JavaScript this would look like:
+```js
+const { position, style } = props
+
+// Or even...
+function MyComponent({
+	position,
+	style,
+})
+```
+
+React properties are themselves an example of a common idiom of passing around large tables as function arguments, such as with HTTP requests:
+
+```js
+// JavaScript
+get("/users", ({
+	users,
+	nextPageCursor,
+}) => { /* code */ })
+```
+
+## Design
+
+This RFC proposes expanding the grammar of `binding`:
+```patch
+binding = NAME [':' Type]
++         | `{` keydestructor [`,` keydestructor] [`,`] `}` [':' Type]
+
++keydestructor = `.` NAME [':' Type]
+```
+
+This would allow for the following:
+
+```lua
+local { .a, .b }, c = t
+
+for _, { .a, .b } in ts do
+end
+
+local function f({ .a, .b }, c)
+end
+```
+
+In all of these cases, `.x` is an index of the table it corresponds to, assigned to a local variable of that name. For example, `local { .a, .b } = t` is syntax sugar for:
+
+```lua
+local a = t.a
+local b = t.b
+```
+
+This will have the same semantics with regards to `__index`, in the order of the variables being assigned. Furthermore, if `t` for whatever reason cannot be indexed (not a table, nil), you will get the same errors as you would if you were writing out `t.a` by hand.
+
+Trying to use object destructuring in a local assignment without a corresponding assignment, such as...
+```lua
+local { .x, .y }
+```
+
+...is valid to parse and will execute, but can be linted against. This might even be emergent, as this will be statically equivalent to `nil.x`, which Luau's type system can catch.
+
+#### Types
+An optional type can be supplied, such as:
+```lua
+local { .a: string, .b: number } = t
+
+-- Equivalent to...
+local a: string = t.a
+local b: number = t.b
+```
+
+Without explicit types, local assignments and for loops will assume the type of `<rhs>.<field>`. For example...
+```lua
+-- x and y will both be typed `number` here
+local { .x, .y } = position :: { number }
+```
+
+As a function argument, it will act the same as if writing it out by hand, where anonymous types are created:
+```lua
+-- Will be f({+ x: a, y: b +})
+local function f({ .x, .y })
+```
+
+Additionally, you can specify a type on the "table" as a whole.
+
+```lua
+local { .x, .y }: Position = p
+```
+
+Combining both is acceptable, in which case the type on the variable takes priority:
+
+```lua
+-- `x` will be `number`, `y` will be the type of `T.y`
+local { .x: number, .y }: T = p
+```
+
+### Non-designs
+#### Renaming variables
+This proposal does not allow something like JavaScript's:
+```js
+const { a: b } = t
+```
+
+...where `t.a` is bound to `b`. This is not undesired, but this RFC is not concerned with it for now. This also means that any field you want to destruct must be a valid Luau identifier.
```

Comment:
Strange that it's not a concern.
I constantly see myself and other people renaming the fields when they use destructuring in C++.

With current proposal, you can't even destruct two objects of the same type in a single scope.

Reply:
> With current proposal, you can't even destruct two objects of the same type in a single scope.

Can you explain what you mean?

Reply:
Do you mean like `const { position: { x, y } } = info()`?

Reply:
Here's one example from Luau code:
```
auto [leftType, leftRefinement] = check(scope, binary->left, relaxedExpectedLhs);
auto [rightType, rightRefinement] = check(scope, binary->right, expectedType);
```

Reply:
Why wouldn't that work in the current proposal?

Reply:
Because local names are forced to be field names.

Reply:
Oh, I see. Yes it is a drag--it is excluded in this RFC out of concern about significantly increasing the scope, while not having it still solves a large amount of use cases.

Reply:
I would not consider C++ structured bindings to be totally isomorphic to the proposed table destructuring, even though it's similar in some ways. C++ structured bindings are _rename by default_, but this means you are forced to have as many binding names as there are fields. With this, you can pick a number of fields (and separately, optionally rename those fields you picked out).

Reply:
fwiw, while field renames are not uncommon in JavaScript, they're pretty rare from what I've seen?

Reply:
But come to think of it, it is important to be able to rename a field in case one shadows the other, which definitely happens all the time.

Reply:
Honestly, I think it might be reasonable to say that this is the good bit of "increasing the scope" for this RFC, and e.g. the argument destructuring was a bad case. It's true that both can be added as a follow-up later, but I think renaming addresses a much more serious need.